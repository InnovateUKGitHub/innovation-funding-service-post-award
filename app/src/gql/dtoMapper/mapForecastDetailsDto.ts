import type { ForecastDetailsDTO } from "@framework/dtos";
import { Clock, salesforceDateFormat } from "@framework/util";

const clock = new Clock();

type ForecastDetailsNode = Readonly<
  Partial<{
    Id: string;
    Acc_CostCategory__c: GQL.Value<string>;
    Acc_LatestForecastCost__c: GQL.Value<number>;
    Acc_ProjectPeriodStartDate__c?: GQL.Value<string>;
    Acc_ProjectPeriodEndDate__c: GQL.Value<string>;
    Acc_ProjectPeriodNumber__c: GQL.Value<number>;
    RecordType: {
      Name: GQL.Value<string>;
    } | null;
  }>
> | null;

type ForecastDetailsDtoMapping = ForecastDetailsDTO;

const mapper: GQL.DtoMapper<ForecastDetailsDtoMapping, ForecastDetailsNode> = {
  id: function (node) {
    return node?.Id ?? "";
  },
  costCategoryId: function (node) {
    return node?.Acc_CostCategory__c?.value ?? "unknown";
  },
  periodId: function (node) {
    return node?.Acc_ProjectPeriodNumber__c?.value ?? 0;
  },
  periodStart: function (node) {
    return clock.parse(node?.Acc_ProjectPeriodStartDate__c?.value ?? null, salesforceDateFormat);
  },
  periodEnd: function (node) {
    return clock.parse(node?.Acc_ProjectPeriodEndDate__c?.value ?? null, salesforceDateFormat);
  },
  value: function (node) {
    return node?.Acc_LatestForecastCost__c?.value ?? 0;
  },
};

/**
 * Maps a specified ForecastDetails Node from a GQL query to a slice of
 * the ForecastDetailsDto to ensure consistency and compatibility in the application
 */
export function mapToForecastDetailsDto<
  T extends ForecastDetailsNode,
  PickList extends keyof ForecastDetailsDtoMapping,
>(node: T, pickList: PickList[]): Pick<ForecastDetailsDtoMapping, PickList> {
  return pickList.reduce((dto, field) => {
    dto[field] = mapper[field](node);
    return dto;
  }, {} as Pick<ForecastDetailsDtoMapping, PickList>);
}

/**
 * Maps ForecastDetails Edge to array of Loan DTOs.
 */
export function mapToForecastDetailsDtoArray<
  T extends ReadonlyArray<{ node: ForecastDetailsNode } | null> | null,
  PickList extends keyof ForecastDetailsDtoMapping,
>(edges: T, pickList: PickList[]): Pick<ForecastDetailsDtoMapping, PickList>[] {
  return (
    edges
      ?.filter(x => x?.node?.RecordType?.Name?.value === "Profile Detail")
      ?.map(node => {
        return mapToForecastDetailsDto(node?.node ?? null, pickList);
      }) ?? []
  );
}
