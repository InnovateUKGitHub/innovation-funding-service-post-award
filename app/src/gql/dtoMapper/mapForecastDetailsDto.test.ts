import { mapToForecastDetailsDtoArray } from "./mapForecastDetailsDto";

describe("mapToForecastDetailsDtoArray", () => {
  const edges = [
    {
      node: {
        Id: "0",
        Acc_CostCategory__c: { value: "Labour" },
        Acc_ProjectPeriodStartDate__c: { value: "2021-01-25" },
        Acc_ProjectPeriodEndDate__c: { value: "2021-03-25" },
        Acc_ProjectPeriodNumber__c: { value: 3 },
        Acc_LatestForecastCost__c: { value: 1000 },
        RecordType: {
          Name: {
            value: "Profile Detail",
          },
        },
      },
    },
    {
      node: {
        Id: "1",
        Acc_CostCategory__c: { value: "Travel" },
        Acc_ProjectPeriodStartDate__c: { value: "2021-01-25" },
        Acc_ProjectPeriodEndDate__c: { value: "2021-03-25" },
        Acc_ProjectPeriodNumber__c: { value: 2 },
        Acc_LatestForecastCost__c: { value: 1000 },
        RecordType: {
          Name: {
            value: "Profile Detail",
          },
        },
      },
    },
    {
      node: {
        Id: "3",
        Acc_CostCategory__c: null,
        Acc_ProjectPeriodStartDate__c: null,
        Acc_ProjectPeriodEndDate__c: null,
        Acc_ProjectPeriodNumber__c: null,
        Acc_LatestForecastCost__c: null,
        RecordType: {
          Name: {
            value: "Profile Detail",
          },
        },
      },
    },
    {
      node: {
        Id: "4",
        Acc_CostCategory__c: { value: "Kickbacks" },
        Acc_ProjectPeriodStartDate__c: { value: "Business" },
        Acc_ProjectPeriodEndDate__c: { value: "CR&D" },
        Acc_ProjectPeriodNumber__c: { value: 7 },
        Acc_LatestForecastCost__c: { value: 1000 },
        RecordType: {
          Name: {
            value: "blah",
          },
        },
      },
    },
  ];

  it("should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records", () => {
    expect(
      mapToForecastDetailsDtoArray(edges, ["id", "costCategoryId", "periodId", "periodStart", "periodEnd", "value"]),
    ).toMatchSnapshot();
  });

  it("should map the gql data only including fields on the picklist", () => {
    expect(mapToForecastDetailsDtoArray(edges, ["id", "periodId", "value"])).toMatchSnapshot();
  });
});
