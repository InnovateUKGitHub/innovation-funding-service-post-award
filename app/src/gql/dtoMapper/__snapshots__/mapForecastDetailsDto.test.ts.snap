// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToForecastDetailsDtoArray should map the gql data only including fields on the picklist 1`] = `
Array [
  Object {
    "id": "0",
    "periodId": 3,
    "value": 1000,
  },
  Object {
    "id": "1",
    "periodId": 2,
    "value": 1000,
  },
  Object {
    "id": "3",
    "periodId": 0,
    "value": 0,
  },
]
`;

exports[`mapToForecastDetailsDtoArray should map the gql data to the correct Dtos filtering out for not New and not Claim Detail records 1`] = `
Array [
  Object {
    "costCategoryId": "Labour",
    "id": "0",
    "periodEnd": 2021-03-25T00:00:00.000Z,
    "periodId": 3,
    "periodStart": 2021-01-25T00:00:00.000Z,
    "value": 1000,
  },
  Object {
    "costCategoryId": "Travel",
    "id": "1",
    "periodEnd": 2021-03-25T00:00:00.000Z,
    "periodId": 2,
    "periodStart": 2021-01-25T00:00:00.000Z,
    "value": 1000,
  },
  Object {
    "costCategoryId": "unknown",
    "id": "3",
    "periodEnd": null,
    "periodId": 0,
    "periodStart": null,
    "value": 0,
  },
]
`;
